pinset/pinset.go

```diff
@@ -0,0 +1,118 @@
+package pinset
+
+import (
+	"fmt"
+
+	"github.com/qri-io/registry"
+)
+
+// ErrPinsetNotSupported is a cannonical error for a repository that does not
+// support pinning
+var ErrPinsetNotSupported = fmt.Errorf("pinset is not supported")
+
+// Pinset is the interface for acting as a remote pinning service.
+// Pinset implementations are expected to keep a store of PinStatus
+// that callers can use to probe the progress of a request
+type Pinset interface {
+	// Pin can take a while, so Pin returns a channel of PinStatus to
+	// deliver updates structs that should all share an ID for the request that
+	// the Pinset implementation will provide
+	Pin(req *PinRequest) (chan PinStatus, error)
+	// Unpin removes a pin
+	Unpin(req *PinRequest) error
+	// Status gets the current pin state value for a given PinRequest
+	Status(req *PinRequest) (PinStatus, error)
+	// Pins lists pins within the range defoined by limit & offset in
+	// lexographical order
+	Pins(limit, offset int) ([]string, error)
+	// PinLen returns the number of pins in the set
+	PinLen() (int, error)
+}
+
+// MemPinset is a completely ficticious implementation of a pinset
+// it shouldn't ever be used in real-world scenarios. We use it for mocking
+// a pinning service without an actual backing store keeping pins
+type MemPinset struct {
+	pk       PinStatusStore
+	Profiles registry.Profiles
+	pins     []string
+}
+
+// Pin a dataset
+func (m *MemPinset) Pin(req *PinRequest) (chan PinStatus, error) {
+	if len(m.pins) == 0 {
+		m.pins = append(m.pins, req.Path)
+	} else {
+		for i, p := range m.pins {
```

Comment:
The intention here is to store the pins in reverse order? This won't work if the Path is less than any other value in the sorted list. For example:

m.pins = ["z", "m", "b"]
Pin({Path: "a"})
During the iteration, req.Path is never > than p, so it'll never be inserted.

Making a function just for insertion, like the one given here, seems better. And it can be directly tested.

Reply:
I just noticed that the Pinset.Pins interface says "lexographical order", which implies this condition is reversed, it should be `req.Path < p`. But my earlier comment still applies, just for inserted an element larger than any other in the list.